app/src/main/java/com/weatherxm/ui/rewardslist/RewardsListViewModel.kt

```diff
@@ -66,66 +56,21 @@ class RewardsListViewModel(
+                usecase.getRewardsTimeline(deviceId, currentPage).map {
                     Timber.d("Got Rewards: ${it.rewards}")
                     hasNextPage = it.hasNextPage
-                    reachedTotal = it.reachedTotal
-                    /*
-                     * There is an edge case where rewards are empty, we have next page
-                     * and we need to hide the spinner (it's literally a success) in the UI,
-                     * but to not refresh the adapter. So we pass null on these occasions.
-                     */
-                    if (it.rewards.isNotEmpty()) {
-                        currentShownRewards.addAll(it.rewards)
-                        onNewRewardsPage.postValue(Resource.success(currentShownRewards))
-                    } else {
-                        onNewRewardsPage.postValue(Resource.success(null))
-                    }
-                }.onLeft {
-                    analytics.trackEventFailure(it.code)
-                }
-
-                blockNewPageRequest = false
-            }
-        } else if (!hasNextPage && !blockNewPageRequest && !reachedTotal) {
-            onNewRewardsPage.postValue(Resource.loading())
-            viewModelScope.launch {
-                currentPage = 0
-                blockNewPageRequest = true
-                currToDate = currFromDate
-                currFromDate = currFromDate.minusMonths(FETCH_INTERVAL_MONTHS)
-
-                usecase.getRewardsTimeline(
-                    deviceId,
-                    currentPage,
-                    currFromDate.toISODate(),
-                    currToDate.toISODate()
-                ).map {
-                    Timber.d("Got Rewards: ${it.rewards}")
-                    hasNextPage = it.hasNextPage
-                    reachedTotal = it.reachedTotal
                     currentShownRewards.addAll(it.rewards)
-                    /*
-                     * There is an edge case where rewards are empty, we have next page
-                     * and we need to hide the spinner (it's literally a success) in the UI,
-                     * but to not refresh the adapter. So we pass null on these occasions.
-                     */
-                    if (it.rewards.isNotEmpty()) {
-                        currentShownRewards.addAll(it.rewards)
-                        onNewRewardsPage.postValue(Resource.success(currentShownRewards))
-                    } else {
-                        onNewRewardsPage.postValue(Resource.success(null))
-                    }
+                    onNewRewardsPage.postValue(Resource.success(currentShownRewards))
                 }.onLeft {
                     analytics.trackEventFailure(it.code)
                 }
+
                 blockNewPageRequest = false
             }
+        } else if (!hasNextPage) {
+            // Add an empty reward to indicate the end of data and send all rewards to the activity
+            currentShownRewards.add(Reward.empty())
```

Comment:
Just a suggestion because I don't like these implicit solutions.
Maybe you can declare a new object with two properties 
`type`: with values `data` and `endOfList`
`data`: `Reward` which will be null if the `type` is `endOfList`

Reply:
Hmm I think that's a valid suggestion. I'm waiting for feedback from @JohnFrontzos (please see Pantelis' suggestion as well, I think his suggestion aligns with yours) before proceeding to implementing this.

Reply:
Yes, I agree with @pantelisss.